Allowfile

skcoulter opened the pull request.

Commit:
input a file containing src and dst lids

Commit:
input a file containing src and dst lids

Commit:
Add ability to input lids via a file

Signed-off-by: Susan K. Coulter <markus@wf-fe1.lanl.gov>

skcoulter commented:
@weiny2 
Ira,
This is my first time committing to your infiniband-diags repo.
I may have screwed it up, the only changes were to ibtracert.c and ibtracert.8.in.
It looks kinda funny.
Lemme know ...

weiny2 reviewed (changes requested):
This should not have been included in the series.  Delete this patch.  The script here is temporary and a build artifact.

weiny2 reviewed (commented):
Did you change the rst file and generate this?  I want to see the rst changes as well.

See: .../infiniband-diags/doc/rst

weiny2 commented on src/ibtracert.c:
```diff
@@ -811,9 +820,29 @@ int main(int argc, char **argv)
+		lid_fd = fopen(lid_file, "r");
+		if (!lid_fd) {
+			IBEXIT("cannot open lid-file %s", lid_file);
+		} else {
```
No need for the else here...  you are exiting if the file does not get opened.

skcoulter replied:
Ah - was not familiar with restructured text and therefore did not see the rst directory.  I made the changes directly in the .in file - thinking it was the "input" to the actual man page.  Will go back and do it right. ( as clearly documented in README.rst :)
Will also redo the code bit that includes the unnecessary else.
You can refuse (or whatever) this one and I'll submit a new one.

The use case is grabbing a "route map" for a large number of nodes - perhaps a job allocation or compute-to-IO, etc.  Calling ibtracert every time is expensive.  As a test, I created a route map for 128 nodes with this patch in just under 3 minutes.  Doing the same thing  by calling ibtracert individually each time took over 21 minutes.

You can see now why Howard was hoping for a friendly maintainer.

weiny2 replied:
Ok.

Could you include this use case in the commit message for the series.

It would also be interesting to evaluate if ibtracert is issuing duplicate queries to the switch tables for each of the S/DLID pairs.  If so that would be another optimization.  [3 minutes still seems like a long time...]

hnrose replied:
It depends on topology of network and number of pairs whose path is being traced. How many hops end to end ?

weiny2 reviewed (commented):
Could you explain the use case of this a bit more?

From the code it seems like you have a single Src/Dest pair of LIDs.  What is the point of being able to feed a single pair in with a file rather than the current command line options?

The patch with the man page says that you can specified a set of LID pairs. How is this done in the code?

Also, squash the C file change with the RST/man page document.  I don't think this needs to be split up that much.

hnrose commented:
@skcoulter 
src/ibtracert should not be part of this. It looks like it's https://github.com/linux-rdma/infiniband-diags/pull/4/commits/27560deb3c17dfd8a5df83e855c165cbbdb60f1b. It is not a checked in file AFAIR so I think this should be removed. I now see Ira wrote this several days ago.

I will go over the code changes as soon as I can.

hnrose commented on src/ibtracert.c:
```diff
@@ -67,7 +67,9 @@ static char *node_type_str[] = {
 static FILE *f;
+static FILE *lid_fd;
 
+static char *lid_file = NULL;
```
parameters could be LIDs or GUIDs.

skcoulter replied:
OK - I'll change name and test with lids or guids

hnrose commented on src/ibtracert.c:
```diff
@@ -811,9 +820,29 @@ int main(int argc, char **argv)
 	argv += optind;
 
-	if (argc < 2)
+	if (argc < 2 && lid_file[0] == '\0') 
```
Should this be:
if (argc < 2 && lid_file == NULL)
?

skcoulter replied:
probably

hnrose commented on src/ibtracert.c:
```diff
@@ -811,9 +820,29 @@ int main(int argc, char **argv)
 
+	if (argc == 2) {
+		srcid = strdup(argv[0]);
+		dstid = strdup(argv[1]);
```
Check for strdup errors and IBEXIT if so

skcoulter replied:
okey dokey

hnrose commented on src/ibtracert.c:
```diff
@@ -811,9 +820,29 @@ int main(int argc, char **argv)
+			IBEXIT("cannot open lid-file %s", lid_file);
+		} else {
+			srcid = (char*) malloc(6*sizeof(char));
+			dstid = (char*) malloc(6*sizeof(char));
```
Check for malloc failures and IBEXIT if so
Also, if GUIDs handled need more than 6 bytes for each malloc

skcoulter replied:
will do

hnrose commented on src/ibtracert.c:
```diff
@@ -811,9 +820,29 @@ int main(int argc, char **argv)
+			dstid = (char*) malloc(6*sizeof(char));
+			lidfd_res = fscanf(lid_fd, "%s %s", srcid, dstid);
+			if (lidfd_res != 2) {
+				IBEXIT("lid-file empty or contains bad data, %i, %s",lidfd_res,lid_file);
```
lidfd_res could be EOF
Would this (when != 2) be better as:
if (lid_res != EOF)
IBEXIT("lid-file (%s) contains bad data %i", lid_file, lidfd_res);
else
IBEXIT("lid-file (%s) empty %s", lid_file, strerror(errno));

skcoulter replied:
hrm - != 2 seems to cover all bases, whether a matching failure or empty/bad file.  lidfs_res of 2 (meaning I got my 2 values) would != EOF, but that is not an error.  Am I missing something here ?

hnrose replied:
It was just an attempt to improve the error message when there wasn't 2 arguments.fscanf says when EOF is returned that errno contains the error so I was trying to separate out that case from the other case.  It's probably overkill.

skcoulter replied:
Yah - I started out with 2 different error messages, but collapsed them cuz it didn't seem necessary. ...I think I'll leave it as is ...

hnrose commented on src/ibtracert.c:
```diff
@@ -825,57 +854,76 @@ int main(int argc, char **argv)
+		if (lid_fd) {
+			my_portid.drpath.cnt = 0;
+			src_portid.drpath.cnt = 0;
+			if (fscanf(lid_fd, "%s %s", srcid, dstid) != 2) {
```
Would be useful to indicate (to stderr) line number in file where error occurred

skcoulter replied:
thx

weiny2 replied:
I'm going to wait for your v2 patch set.  Please include the main use case of the patch in the commit and doc.  Thanks.

hnrose reviewed (changes requested):
In general, this is a good idea.

I think it can handle both GUIDs and LIDs as parameters in the file.

Nit: Also, should follow code style used.

Commit:
Allow specificaiton of a file that contains several src/dst pairs.
These can be lids or guids, in non-sequential order.
The use case is allowing the creation of large route maps in a significantly shorter period of time.
Testing a route map for 128 ports, ((2*128)**2 - 2n) = 32512 , resulted in over 7x speed up.

Signed-off-by: Susan K. Coulter <markus@wf-fe1.lanl.gov>

Conflicts:

	src/ibtracert.c

Signed-off-by: Susan K. Coulter <markus@wf-fe1.lanl.gov>

skcoulter commented:
@weiny2 check this one out

skcoulter commented:
@weiny2 - hold the phone, crap ... stand by

Commit:
conflict fix needed a fix

Signed-off-by: Susan K. Coulter <markus@wf-fe1.lanl.gov>

skcoulter commented:
@weiny2 OK - newbie git error with mixing branches caused conflicts that I manually fixed - but not quite the first time ... now should be good.  It actually compiles.  :)  And testing succeeds too.

hnrose commented on src/ibtracert.c:
```diff
@@ -759,6 +761,13 @@ static int process_opt(void *context, int ch, char *optarg)
 	switch (ch) {
 	case 1:
 		node_name_map_file = strdup(optarg);
+		if (node_name_map_file == 0)
```
NULL r.t. 0 ?

hnrose commented on src/ibtracert.c:
```diff
@@ -759,6 +761,13 @@ static int process_opt(void *context, int ch, char *optarg)
+		break;
+	case 2:
+		ports_file = strdup(optarg);
+		if (ports_file == 0)
```
NULL r.t. 0 ?

hnrose commented on src/ibtracert.c:
```diff
@@ -811,9 +826,35 @@ int main(int argc, char **argv)
 
+	if (argc == 2) {
+		srcid = strdup(argv[0]);
+		if (srcid == 0)
```
NULL r.t. 0

hnrose commented on src/ibtracert.c:
```diff
@@ -811,9 +826,35 @@ int main(int argc, char **argv)
+		if (srcid == 0)
+			IBEXIT("out of memory, strdup for srcid failed");
+		dstid = strdup(argv[1]);
+		if (dstid == 0)
```
NULL r.t. 0

hnrose commented on doc/man/ibtracert.8.in:
```diff
@@ -1,8 +1,8 @@
-.\" Man page generated from reStructuredText.
```
Shouldn't this be removed as it's generated file ?

skcoulter replied:
Yes - but I don't know how.  Suggestions from stackoverflow look really scary.

hnrose commented on src/ibtracert:
```diff
@@ -0,0 +1,148 @@
+#! /bin/sh
```
Shouldn't this be removed as it's generated file ?

hnrose commented on src/ibtracert.c:
```diff
@@ -825,57 +866,83 @@ int main(int argc, char **argv)
+		if (resolve_portid_str(ibd_ca, ibd_ca_port, &src_portid, srcid,
 			       ibd_dest_type, ibd_sm_id, srcport) < 0)
-		IBEXIT("can't resolve source port %s", argv[0]);
+			IBEXIT("can't resolve source port %s", argv[0]);
```
argv[0] or srcid ?

hnrose commented on src/ibtracert.c:
```diff
@@ -825,57 +866,83 @@ int main(int argc, char **argv)
-			IBEXIT("cannot resolve lid for port \'%s\'",
-				portid2str(&dest_portid));
-	}
+			IBEXIT("can't resolve destination port %s", argv[1]);
```
argv[1] or dstid ?

hnrose commented on src/ibtracert.c:
```diff
@@ -825,57 +866,83 @@ int main(int argc, char **argv)
+		if (ports_fd) {
+			my_portid.drpath.cnt = 0;
+			src_portid.drpath.cnt = 0;
+     	   		if (fgets(ports_buffer, sizeof(ports_buffer), ports_fd)) {
```
Indentation looks wrong here. One too many tabs ?

hnrose commented on src/ibtracert.c:
```diff
@@ -825,57 +866,83 @@ int main(int argc, char **argv)
+			src_portid.drpath.cnt = 0;
+     	   		if (fgets(ports_buffer, sizeof(ports_buffer), ports_fd)) {
+				num_port_pairs++;
+                		if ( sscanf(ports_buffer, "%s %s", srcid, dstid) != 2 ) 
```
Indentation looks wrong here. One too many tabs ?

hnrose commented on src/ibtracert.c:
```diff
@@ -790,6 +804,7 @@ int main(int argc, char **argv)
 		{"no_info", 'n', 0, NULL, "simple format"},
 		{"mlid", 'm', 1, "<mlid>", "multicast trace of the mlid"},
 		{"node-name-map", 1, 1, "<file>", "node name map file"},
+		{"ports-file", 2, 1, "<file>", "ports file"},
```
Rather than just "ports file" as last parameter, maybe something like "ports file containing either LID or GUID pairs"

hnrose reviewed (changes requested):
Looks mostly good to me ;-)

Mainly nits

skcoulter commented:
All suggested changes made except the indentation, which looks right to me.  I don't know how to remove the generated files from this bundle of goo, so I'll wait to commit until I figure that out or get more guidance from you.

weiny2 commented:
I have been checking in generated files in case a developer does not have the tools to make them.  The build system should check for that.  So including the .in _and_ the .rst is the correct thing to do.  The key is that the diff should be on the .rst _first_ and the .in is just a build result.

I’ll check the patch soon.  Did you update after Hal’s comments?

Ira


From: Susan Coulter [mailto:notifications@github.com]
Sent: Friday, February 16, 2018 9:54 AM
To: linux-rdma/infiniband-diags <infiniband-diags@noreply.github.com>
Cc: Weiny, Ira <ira.weiny@intel.com>; Mention <mention@noreply.github.com>
Subject: Re: [linux-rdma/infiniband-diags] Allowfile (#4)


All suggested changes made except the indentation, which looks right to me. I don't know how to remove the generated files from this bundle of goo, so I'll wait to commit until I figure that out or get more guidance from you.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/linux-rdma/infiniband-diags/pull/4#issuecomment-366309531>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AACVlf_D3JGLoR_HEFdHao6ajhmdEMTQks5tVcCigaJpZM4R0gF6>.

Commit:
Changes as requested by Hal.

Signed-off-by: Susan K. Coulter <markus@wf-fe1.lanl.gov>

skcoulter commented:
On Feb 16, 2018, at 1:04 PM, Ira Weiny <notifications@github.com<mailto:notifications@github.com>> wrote:

I have been checking in generated files in case a developer does not have the tools to make them. The build system should check for that. So including the .in _and_ the .rst is the correct thing to do. The key is that the diff should be on the .rst _first_ and the .in is just a build result.

I’ll check the patch soon. Did you update after Hal’s comments?

Yessir.
Except the indentation stuff looked right to me - so nothing was done wrt those comments.
The other changes were all done, and just now committed.


Ira


From: Susan Coulter [mailto:notifications@github.com]
Sent: Friday, February 16, 2018 9:54 AM
To: linux-rdma/infiniband-diags <infiniband-diags@noreply.github.com<mailto:infiniband-diags@noreply.github.com>>
Cc: Weiny, Ira <ira.weiny@intel.com<mailto:ira.weiny@intel.com>>; Mention <mention@noreply.github.com<mailto:mention@noreply.github.com>>
Subject: Re: [linux-rdma/infiniband-diags] Allowfile (#4)


All suggested changes made except the indentation, which looks right to me. I don't know how to remove the generated files from this bundle of goo, so I'll wait to commit until I figure that out or get more guidance from you.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/linux-rdma/infiniband-diags/pull/4#issuecomment-366309531>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AACVlf_D3JGLoR_HEFdHao6ajhmdEMTQks5tVcCigaJpZM4R0gF6>.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/linux-rdma/infiniband-diags/pull/4#issuecomment-366343866>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABFqvML7EqjUh9JVXuNdGJGBy7vvaPBRks5tVd9FgaJpZM4R0gF6>.

================================
Susan Coulter
LANL / USRC / HPC-DES
Network Lead
(505) 412-6525
Lokah Samastah Sukhino Bhavantu
================================
